src/components/main-form.js

```diff
@@ -1,180 +1,192 @@
 import { FV, PMT } from 'formulajs/lib/financial';
 
+
+const COMPOUND_FREQUENCY = 12;
+
+const o = {
+  interestAmt: 0,
+  interestRate: 0,
+  term: 0,
+  pmt: 0,
+  fv: 0
+};
+
+let o1 = Object.assign({}, o);
+let o2 = Object.assign({}, o);
+
+o1.interestRate = 4.3;
+o1.term = 30;
+
+o2.interestRate = 4;
+o2.term = 15;
+
+const Row = styled('div')`
+  display: flex;
+  flex-direction: row;
+`
+
 class MainForm extends React.Component {
 
   constructor() {
     super();
-
-    this.calculate = this.calculate.bind(this);
-    this.updateInput = this.updateInput.bind(this);
-    this.validateData = this.validateData.bind(this);
-
-    this.state = this.calculate({
-      investmentRate: 7,
-      mortgageAmt: 200000,
+    this.calculatePMT = this.calculatePMT.bind(this)
+    this.calculateInterestAmt = this.calculateInterestAmt.bind(this)
+    this.calculateFV = this.calculateFV.bind(this)
+    this.calculateOptCost = this.calculateOptCost.bind(this);
+    this.updateInput = this.updateInput.bind(this)
+    this.calculateAll = this.calculateAll.bind(this);
+    this.state = {
+      investmentRate: 9,
+      loanAmt: 200000,
+      inflation: 2,
+      optCost: 0,
       options: {
-        a: {
-          fv: 0,
-          interestAmt: 0,
-          mortgageInterestRate: 4.3,
-          mortgageTerm: 30,
-          pmt: 0,
-        },
-        b: {
-          fv: 0,
-          interestAmt: 0,
-          mortgageInterestRate: 4,
-          mortgageTerm: 15,
-          pmt: 0,
-        }
+        o1, o2
       }
-    });
+    }
   }
 
-  updateInput(e) {
-    let state = _.cloneDeep(this.state);
-    const key = e.target.name;
+  componentWillMount() {
+    this.calculateAll(this.state);
+  }
 
-    // key and value be explicit
+  calculatePMT(o, state) {
+    return PMT((o.interestRate - state.inflation) / 100 / COMPOUND_FREQUENCY, o.term * COMPOUND_FREQUENCY, state.loanAmt);
+  }
 
-    const value = parseFloat(e.target.value);
+  calculateInterestAmt(o, state) {
+    return o.pmt * (o.term * COMPOUND_FREQUENCY) + state.loanAmt;
+  }
 
-    _.set(state, key, value);
-    if (!isNaN(value)) {
-      state = this.calculate(state);
+  calculateFV(o1, o2, state) {
+    // TODO: what if o1 and o2 are equal?
+    if (o1.term > o2.term) {
+      return FV((state.investmentRate - state.inflation) / 100 / COMPOUND_FREQUENCY, o1.term * COMPOUND_FREQUENCY, o2.pmt - o1.pmt, 0);
+    } else if (o1.term < o2.term) {
+      return FV((state.investmentRate - state.inflation) / 100 / COMPOUND_FREQUENCY, (o2.term - o1.term) * COMPOUND_FREQUENCY, o1.pmt, 0);
     }
-    this.setState(state);
   }
 
-  validateData(obj) {
-    for (let key in obj)
-      if (typeof obj[key] === 'object') {
-        return this.validateData(obj[key])
-      }
-      else if (!obj[key] && obj[key] !== 0) {
-        return false;
-      }
-    return true;
+  calculateOptCost(o1, o2) {
+    return (o1.fv + o1.interestAmt) - (o2.fv + o2.interestAmt);
   }
```

Comment:
sug: Could you document these methods so that we know what they are for in the future?

Reply:
of course